compiler/rustc_hir_analysis/src/coherence/orphan.rs

```diff
@@ -404,23 +407,74 @@ fn emit_orphan_check_error<'tcx>(
+
+            let mut reported = None;
+            for (param, span) in collector.uncovered_params {
+                reported.get_or_insert(match local_type {
+                    Some(local_type) => tcx.dcx().emit_err(errors::TyParamFirstLocal {
+                        span,
+                        note: (),
+                        param,
+                        local_type,
+                    }),
+                    None => tcx.dcx().emit_err(errors::TyParamSome { span, note: (), param }),
+                });
+            }
+
+            struct UncoveredTyParamCollector<'tcx> {
+                tcx: TyCtxt<'tcx>,
+                infcx: Option<Box<InferCtxt<'tcx>>>,
+                generics: &'tcx ty::Generics,
+                uncovered_params: FxIndexMap<Symbol, Span>,
+            }
+
+            impl<'tcx> UncoveredTyParamCollector<'tcx> {
+                fn add_uncovered_param_ty(&mut self, index: u32) {
+                    let param_def = self.generics.param_at(index as _, self.tcx);
+                    let span = self.tcx.def_ident_span(param_def.def_id).unwrap();
+                    self.uncovered_params.insert(param_def.name, span);
                 }
             }
 
-            match local_type {
-                Some(local_type) => tcx.dcx().emit_err(errors::TyParamFirstLocal {
-                    span: sp,
-                    note: (),
-                    param_ty,
-                    local_type,
-                }),
-                None => tcx.dcx().emit_err(errors::TyParamSome { span: sp, note: (), param_ty }),
+            impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for UncoveredTyParamCollector<'tcx> {
+                type BreakTy = !;
+
+                fn visit_ty(&mut self, ty: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+                    match *ty.kind() {
+                        ty::Param(param_ty) => {
+                            self.add_uncovered_param_ty(param_ty.index);
+                            return ControlFlow::Continue(());
+                        }
+                        ty::Infer(ty::TyVar(_)) => {
+                            if let Some(infcx) = &self.infcx
+                                && let Some(origin) = infcx.type_var_origin(ty)
+                                && let TypeVariableOriginKind::TypeParameterDefinition(_, def_id) =
+                                    origin.kind
+                                && let Some(index) =
+                                    self.generics.param_def_id_to_index(self.tcx, def_id)
+                            {
+                                self.add_uncovered_param_ty(index);
+                                return ControlFlow::Continue(());
+                            }
+                        }
+                        _ => {}
+                    }
+
+                    ty.super_visit_with(self)
+                }
             }
+
+            reported.unwrap_or_else(|| bug!("failed to find ty param in `{uncovered_ty}`"))
```

Comment:
I think this is reachable :thinking: if we replace a param with an infer, and then normalize e.g. `<() as Trait<T>>::Assoc` (`<() as Trait<?t>>::Assoc` by using an `impl<U> Trait<(U,)> for () { type Assoc = U; }`. This results in an uncovered infer var which represents `?u` after instantiating `?t` with `(?u,)` :thinking: 

don't have a strong opinion on how to handle this :shrug:

Reply:
I think the easiest approach is to not map back and simply have a worse error message here, maybe adding a fixme :grin:

Reply:
@lcnr I'm failing to find a reproducer of this form. Sure, this projection normalizes to a type var `?proj` that's distinct from the “input type var” `?t` but since we're falling back to the user type `ty` in `OrphanChecker::visit_ty` if the normalized type `norm_ty` `is_ty_var`, the diagnostic code in `orphan.rs` only has to deal with `?t` which it can remap successfully.

I'm not saying that this `bug!()` will never get reached, it does look very dubious. However, I couldn't come up with an example of that yet.